### ISSUE TITLE
Investigate "Profiler disabled because there was a problem in setting up the libddprof dependency"

### ISSUE THREAD

**Issue by ivoanjo** [DataDog/dd-trace-rb#2068]
This was reported by @rgrey in https://github.com/DataDog/dd-trace-rb/issues/2045#issuecomment-1147329082 , and I've extracted it to a separate issue so it can be investigated separately.

The issue is that profiling does not work with the following error message:

> WARN -- ddtrace: [ddtrace] Profiling was requested but is not supported, profiling disabled: Your ddtrace installation is missing support for the Continuous Profiler because there was a problem in setting up the libddprof dependency. For help solving this issue, please contact Datadog support at https://docs.datadoghq.com/help/.

I'll add the next steps for this investigation soon.

**Comment by ivoanjo** [comment]
I've been looking into this issue, but discovered that there's a bug/limitation in rubygems where the log file we would need to debug this issue is overwritten because dd-trace-rb ships two native extensions that both need compiling.

@rgrey Would you be open to trying a development build of ddtrace that should be able to log more information about why the failure is happening? (If not, please do let me know, I can look into speeding up a release with this extended logging)

You can get it from <https://s3.amazonaws.com/gems.datadoghq.com/prerelease-v2/gems/ddtrace-1.1.0.ivoanjo.improve.native.extension.debugging.232295.gem> and here's how to install it:

`$ DD_PROFILING_FAIL_INSTALL_IF_MISSING_EXTENSION=true gem install ddtrace-1.1.0.ivoanjo.improve.native.extension.debugging.232295.gem`

I've added the `DD_PROFILING_FAIL_INSTALL_IF_MISSING_EXTENSION` environment variable in #2069 and it should make it so that the error you're seeing happen immediately at installation-time, and Ruby will give you the path to the `mkmf.log` file that gets generated with more details on why installation failed.

**Comment by rgrey** [comment]
```
LD_LIBRARY_PATH=.:/usr/local/lib "gcc -o conftest -I/usr/local/include/ruby-3.1.0/x86_64-linux -I/usr/local/include/ruby-3.1.0/ruby/backward -I/usr/local/include/ruby-3.1.0 -I.    -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef  -fPIC conftest.c  -L. -L/usr/local/lib -Wl,-rpath,/usr/local/lib -L. -fstack-protector-strong -rdynamic -Wl,-export-dynamic -Wl,--no-as-needed     -Wl,-rpath,/usr/local/lib -L/usr/local/lib -lruby  -lm  -lc"
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

LD_LIBRARY_PATH=.:/usr/local/lib "gcc -I/usr/local/include/ruby-3.1.0/x86_64-linux -I/usr/local/include/ruby-3.1.0/ruby/backward -I/usr/local/include/ruby-3.1.0 -I.    -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef  -fPIC  -std=gnu99 -Werror -c conftest.c"
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

LD_LIBRARY_PATH=.:/usr/local/lib "gcc -I/usr/local/include/ruby-3.1.0/x86_64-linux -I/usr/local/include/ruby-3.1.0/ruby/backward -I/usr/local/include/ruby-3.1.0 -I.    -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef  -fPIC -std=gnu99  -Wno-unused-function -Werror -c conftest.c"
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

LD_LIBRARY_PATH=.:/usr/local/lib "gcc -I/usr/local/include/ruby-3.1.0/x86_64-linux -I/usr/local/include/ruby-3.1.0/ruby/backward -I/usr/local/include/ruby-3.1.0 -I.    -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef  -fPIC -std=gnu99 -Wno-unused-function  -Wno-declaration-after-statement -Werror -c conftest.c"
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

LD_LIBRARY_PATH=.:/usr/local/lib "gcc -I/usr/local/include/ruby-3.1.0/x86_64-linux -I/usr/local/include/ruby-3.1.0/ruby/backward -I/usr/local/include/ruby-3.1.0 -I.    -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef  -fPIC -std=gnu99 -Wno-unused-function -Wno-declaration-after-statement  -Werror-implicit-function-declaration -Werror -c conftest.c"
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

LD_LIBRARY_PATH=.:/usr/local/lib "gcc -I/usr/local/include/ruby-3.1.0/x86_64-linux -I/usr/local/include/ruby-3.1.0/ruby/backward -I/usr/local/include/ruby-3.1.0 -I.    -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef  -fPIC -std=gnu99 -Wno-unused-function -Wno-declaration-after-statement -Werror-implicit-function-declaration  -fvisibility=hidden -Werror -c conftest.c"
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

 [ddtrace] PKG_CONFIG_PATH set to ':/usr/local/bundle/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux/libddprof-x86_64-unknown-linux-gnu/lib/pkgconfig'
package configuration for ddprof_ffi_with_rpath is not found
 [ddtrace] Failure cause: there was a problem in setting up the `libddprof` dependency. For help solving this issue, please contact Datadog support at <https://docs.datadoghq.com/help/>.
GET | 200 | 233 ms | envoy
```

**Comment by ivoanjo** [comment]
Thanks for the log! Just to add some context, `mkmf` (that generates `mkmf.log`) is a tool that's part of the Ruby standard library and get gets called when compiling native extensions, like ddtrace requires.

When a Ruby native extension wants to compile and link to an external native library, `mkmf` invokes a tool called `pkg-config` to get the correct settings to do this. This is the step that is failing here -- `package configuration for ddprof_ffi_with_rpath is not found` is actually logged directly by Ruby's `mkmf`.

Could you share the results of running the following commands:

* :one: `pkg-config --version`
* :two: `PKG_CONFIG_PATH=:/usr/local/bundle/gems/libddprof-0.6.0.1.0-x86_64-linux/vendor/libddprof-0.6.0/x86_64-linux/libddprof-x86_64-unknown-linux-gnu/lib/pkgconfig pkg-config --debug ddprof_ffi_with_rpath`

I suspect these two final clues will allow me to get to the bottom of this issue and produce a fix for your situation. Thanks for the patience so far!

**Comment by rgrey** [comment]
```
$ pkg-config --version
bash: pkg-config: command not found
```
It appears to build fine if we install pkg-config 😁

**Comment by ivoanjo** [comment]
👋 @rgrey the improved message has been released in [v1.2.0](https://github.com/DataDog/dd-trace-rb/releases/tag/v1.2.0).
No more users shall be confused by this issue :)
